### PR TITLE
Fix in the procedures when timer Ta fires

### PR DESCRIPTION
**Pull request by cdh4u**

### CODE REVIEW

**Review by taylor-b — COMMENTED** [review]
This looks correct, but it would be nice if it could be simplified. There's a lot of duplication between the first two and last two paragraphs. Is there a way the algorithm could be structured like this?

```
if triggered queue non-empty:
  send triggered check
else:
  if no pair is Waiting:
    for each foundation:
      if no pair is waiting or in-progress:
        unfreeze first pair
  send check from first Waiting pair
```

**Comment by cdh4u** [comment]
I agree it would be good if we could simplify the text. I have been thinking about how it could be done, e.g., using a table, a flow chart, and/or pseudo code (per your example).

**Comment by cdh4u** [comment]
I added a flow chart that illustrates the procedures. Hopefully it makes things easier to understand :)

File: draft-ietf-ice-rfc5245bis.xml [hunk 1]
```diff
@@ -2067,6 +2067,94 @@ the Waiting state, performs a connectivity check on that pair and puts the
...
+|associated with selected  |------->|CHECK QUEUE, set the pair |   |
+|CHECK LIST?               |        |state to In-Progress and  |   |
+|                          |        |perform connectivity check|   |
+|                          |        |on the pair. Restart Ta.  |   |
++--------------------------+        +==========================+   |
+             |                                                     |
+             | NO                                                  |
+             v                                                     |
++--------------------------+                                       |
+|Candidate Pair(s) in      |   YES                                 |
```
Comment (taylor-b) [inline draft-ietf-ice-rfc5245bis.xml:2100]: For "YES", this looks like it goes back to "select a check list", creating an endless loop?

File: draft-ietf-ice-rfc5245bis.xml [hunk 2]
```diff
@@ -2067,6 +2067,94 @@ the Waiting state, performs a connectivity check on that pair and puts the
...
+|                          |------->|candidate pair in Waiting |   |
+|                          |        |state and set the pair    |   |
+|                          |        |state to In-Progress.     |   |
+|                          |        |Restart Ta.               |   |
++--------------------------+        +==========================+   |
+             |                                                     |
+             | NO                                                  |
+             v                                                     |
++--------------------------+                                       |
+|Candidate Pair(s) in      |   NO                                  |
```
Comment (taylor-b) [inline draft-ietf-ice-rfc5245bis.xml:2119]: Same here, for "NO".

**Review by taylor-b — COMMENTED** [review]
The flow chart may be a helpful addition, but as for simplifying the normative text itself, I was thinking something more along these lines: https://github.com/ice-wg/rfc5245bis/pull/34

File: draft-ietf-ice-rfc5245bis.xml (hunk 1 above)
Comment (cdh4u) [reply draft-ietf-ice-rfc5245bis.xml:2100]: I meant to say checking the next check list in the check list set.

File: draft-ietf-ice-rfc5245bis.xml (hunk 2 above)
Comment (cdh4u) [reply draft-ietf-ice-rfc5245bis.xml:2119]: See above.

**Comment by cdh4u** [comment]
I was thinking. Shouldn't one first, BEFORE checking the TRIGGERED CHEQUE LIST, check whether there is a candidate pair in the In-Progress state?

**Comment by taylor-b** [comment]
You mean so that the time slot will be used to retransmit a connectivity check for the In-Progress pair, rather than sending a new check? Yes, that makes sense.

Also, did you have any opinions about PR #34?

**Comment by cdh4u** [comment]
>You mean so that the time slot will be used to retransmit a connectivity check for the In-Progress pair, >rather than sending a new check? Yes, that makes sense.

Correct.

>Also, did you have any opinions about PR #34?

Had a quick look, but will take a closer look in the office tomorrow.

**Pull request closed without merging**